src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -37,6 +37,38 @@ internal class DiagnosticComputer
         /// pressure when there are large number of open documents across different projects to be analyzed by background analysis.
         /// </summary>
         private static CompilationWithAnalyzersCacheEntry? s_compilationWithAnalyzersCache = null;
+
+        /// <summary>
+        /// List of high priority diagnostic computation tasks which are currently executing.
+        /// Any new high priority diagnostic request is added to this list before the core diagnostics
+        /// compute call is performed, and removed from this list after the computation finishes.
+        /// Any new normal priority diagnostic request first waits for all the high priority tasks in this list
+        /// to complete, and moves ahead only after this list becomes empty.
+        /// </summary>
+        /// <remarks>
+        /// Read/write access to the list is guarded by <see cref="s_gate"/>.
+        /// </remarks>
+        private static readonly List<Task> s_highPriorityComputeTasks = new();
+
+        /// <summary>
+        /// List of cancellation token sources for normal priority diagnostic computation tasks which are currently executing.
+        /// For any new normal priority diagnostic request, a new cancellation token source is created and added to this list
+        /// before the core diagnostics compute call is performed, and removed from this list after the computation finishes.
+        /// Any new high priority diagnostic request first fires cancellation on all the cancellation token sources in this list
+        /// to avoid resource contention between normal and high priority requests.
+        /// Canceled normal priority diagnostic requests are re-attempted from scratch after all the high priority requests complete.
+        /// </summary>
+        /// <remarks>
+        /// Read/write access to the list is guarded by <see cref="s_gate"/>.
+        /// </remarks>
+        private static readonly List<CancellationTokenSource> s_cancellationSourcesForNormalPriorityComputeTasks = new();
```

Comment:
note to self: can this just be a CancellationSeries?

Reply:
Possibly, but I'll stay with the current implementation for simplicity for now.